doc/designs/quic-design/quic-record-layer.md

```diff
@@ -0,0 +1,353 @@
+QUIC Record Layer API
+=====================
+
+The QUIC record layer (QRL) performs the following functions:
+
+- The handling of received datagrams and the initial processing of the
+  packets within them, including packet header processing and packet
+  decryption.
+
+- The serialization, encryption and transmission of sent packets.
+
+The QRL relies on the DEMUX, which is responsible for actual reads of datagrams
+from a datagram network BIO. The DEMUX performs a preliminary examination of
+received packet headers to determine a destination connection ID and passes
+received datagrams to a specific instance of a QUIC record layer based on the
+destination connection ID.
+
+The QRL does not process or examine decrypted packet payloads Processing of
+frames is performed by other components, which read them from the QRL.
+
+The QRL does not use the same method structures or polymorphic interface as the
+TLS record layer, as there is no benefit to doing so and the structure of the
+records returned differs substantially.
+
+Broadly, the interface to the QRL looks like this:
+
+Instantiation
+-------------
+
+```c
+#define OSSL_QUIC_ENC_LEVEL_INITIAL     0
+#define OSSL_QUIC_ENC_LEVEL_HANDSHAKE   1
+#define OSSL_QUIC_ENC_LEVEL_0RTT        2
+#define OSSL_QUIC_ENC_LEVEL_1RTT        3
+
+typedef struct ossl_qrl_st OSSL_QRL;
+
+typedef struct ossl_qrl_args_st {
+    OSSL_LIB_CTX *libctx;
+    const char   *propq;
+
+    /*
+     * This list is not exhaustive and will expand as the QRL
+     * is developed and its requirements become more clear.
+     */
+
+    /* Datagram network BIO to use. */
+    BIO    *rx_bio, *tx_bio;
+    /* Length of connection IDs in short packets in bytes. */
+    size_t  short_conn_id_len;
+    /* key ... */
+    /* IV ... */
+    /* cipher information ... */
+
+} OSSL_QRL_ARGS;
+
+/*
+ * Instantiates a new QRL. A pointer to the QRL is written
+ * to *qrl. Returns 1 on success or 0 on failure.
+ */
+int ossl_qrl_new(const OSSL_QRL_ARGS *args, OSSL_QRL **qrl);
+
+/*
+ * Frees the QRL. All packets obtained using ossl_qrl_read_pkt must already have
+ * been released by calling ossl_qrl_release_pkt.
+ */
+int ossl_qrl_free(OSSL_QRL *qrl);
+```
+
+RX API
+------
+
+```c
+/*
+ * Information about a received QUIC packet.
+ */
+typedef struct ossl_qrl_rx_pkt_st {
+    /* Opaque handle to be passed to ossl_qrl_release_pkt. */
+    void               *handle;
+
+    /* Points to a logical decode of the QUIC packet header. */
+    OSSL_QUIC_PKT_HDR  *hdr;
```

Comment:
Is there a reason this has to be a pointer?  There will always be a unique `OSSL_QUIC_PKT_HDR` for each `OSSL_QRL_RX_PKT`, right?  So why not just:

``` C
    OSSL_QUIC_PKT_HDR hdr;
```

Reply:
The same question applies to `OSSL_QRL_TX_PKT` below, of course

Reply:
The OSSL_QRL_RX_PKT structure is copied on the call to read_pkt. My expectation is that the QRL will probably process potentially multiple packets in advance (and therefore have already decoded the header) and have them waiting to read. Using a pointer avoids copying the OSSL_QUIC_PKT_HDR structure.

Reply:
Not sure I understand...  *Each packet has a header of its own*, right?  So how several packets would share the same `OSSL_QRL_PKT_HDR`, I don't quite understand...

Reply:
They don't share a header. When the packet header is decoded into a `OSSL_QUIC_PKT_HDR`, `ossl_qrl_read_pkts` has not been called yet. So by the time `ossl_qrl_read_pkts` is called, the `OSSL_QUIC_PKT_HDR` has already been written to somewhere else internally inside the QRL. So we can either provide a pointer to the already-decoded internal structure or copy from it.

Reply:
Ok.